Add API for setting clipboard contents

psychon opened the pull request:
This fixes #2642. (although I was too lazy to implement `MULTIPLE` and will just wait for someone to request this; even though the spec requires support for it, I still doubt it is used much. We will see...)

Example code that I used for testing:
```lua
    local o1 = selection_acquire("PRIMARY")
    o1:connect_signal("release", function() print("o1 released") end)
    o1:connect_signal("request", function(_, target, t)
        if target == "TARGETS" then
            t:send{ data = { "TARGETS", "UTF8_STRING" }, format = "atom" }
        else
            local data = {}
            for i=1, 2^23 do
                table.insert(data, "Hi")
            end
            print(#data)
            local data = table.concat(data)
            print(#data)
            print(string.format("data has length %d == 0x%x, %d bytes of space left", #data, #data, 16777212-#data))
            t:send{ data = data, continue = true }

            local count = 0
            t:connect_signal("continue", function()
                print("got continue signal");
                count = count + 1
                if count == 0 then
                    t:send{ data = ". Thanks for listening.", continue = true }
                elseif count == 1 then
                    t:send{ data = "", continue = true }
                elseif count == 2 then
                    t:send{ data = "\n" }
                end
            end)
        end
    end)
```

Fun fact: I looked at the GTK API after writing this and the API that I came up with is surprisingly close to GTK's API. Isn't GTK's API known for being hard to use...?

Elv13 commented on objects/selection_acquire.c:
```diff
@@ -0,0 +1,240 @@
+    xcb_atom_t name_atom;
+    selection_acquire_t *selection;
+
+    name = luaL_checklstring(L, 2, &name_length);
```
For consistency and futureproofness considerations, can we have an `args` table even if there's (currently) only a single argument?

Elv13 commented:
Thanks! There's a Codacy warning, but beside :+1:

psychon commented:
> The scope of the variable 'data' can be reduced.

Aha... -.-
Updated to fix that and that dismissed your review. No force-push, I just added another commit to this PR.

blueyed commented on tests/test-selection-transfer.lua:
```diff
@@ -0,0 +1,235 @@
+    wait_a_bit,
+    wait_a_bit,
+    wait_a_bit,
+    wait_a_bit,
```
Why are those `wait_a_bit` calls necessary?  Could have a comment maybe, if necessary.

psychon replied:
Done in a new commit.

Because transferring 100 MiB of data (inefficiently) is slow. I had a timeout without `wait_a_bit` (which means that it took longer than 0.5 seconds, I think). I just added a couple of these calls because others might have slower computers than I do.

The previous step tests large data transfer. The goal here is to have a single piece of data that is so large that the C code has to split it up into pieces, too. Thus, a single `:send` turns into multiple `INCR` steps for the C code. This also tests continuation, i.e. Lua adding more data later. If my math is correct, this almost transfers 100 MiB of data through the X11 server.

The maximum request size of my current X11 server is a bit below 16 MiB. Thus, each of the 32 MiB pieces that Lua gives to C has to split up into three increments, two almost of full size and one with just 8 bytes of data. :-)

Elv13 commented:
Last call. If there is no other reviews I will merge it manually tomorrow. This is blocking two pull requests.

@blueyed Can you draft the public API you want for this? I think I get what kind of use cases you are after, but if you can actually document them, it would be useful. So far you mentioned:

 * Using rules like awful.rules to define when to transfer ownership
 * Having some ways to running these rules both when the seletion changes and "at any time" like before SIGSTOPping clients

And me:

 * Being able to store and keep "n" selection from the history for some payload types
 * Have an awful.widget.common widget to list them
 * Being able to restore a recent selection from history
 * Move a selection from the PRIMARY and SELECTION buffers using a keybinding

This leaves the problem of knowing which client the selection comes from (if any, ei. `xsel`) unsolved. But it isn't really important for my use case and apparently isn't possible to implement in a reliable way (like assuming the focused client is the source, traveling the winID tree until WM_NAME is set or using some LD_PRELOAD magic to set an extra xproperty on it when a client uses `awful.spawn` to be launched)

blueyed commented on objects/selection_acquire.c:
```diff
@@ -0,0 +1,243 @@
+            {
+                /* Remove table and key */
+                lua_remove(L, -2);
+                lua_remove(L, -2);
```
I am confused by `lua_remove` using `-2` here twice.. would have expected it to be -2 and -1.

psychon replied:
Well... the stack at this point (from bottom to top):

- The table that we are iterating over (Index -3)
- The current key in the table (Index -2)
- The current value in the table (Index -1)

This wants to get rid of the table and the key, but wants to preserve the value. Thus, the elements at stack index -2 and -3 should be removed. The code here first removes the key, so afterwards the table is at index -2, i.e. below the value.

Does this make it clearer?

blueyed replied:
Yes, thanks!

blueyed commented:
> Last call. If there is no other reviews I will merge it manually tomorrow. This is blocking two pull requests.

Left a hopefully minor comment.
In general I cannot say to much about it, feel free to merge after confirming that the `-2` is correct.

blueyed commented:
> @blueyed Can you draft the public API you want for this? I think I get what kind of use cases you are after, but if you can actually document them, it would be useful. So far you mentioned:
> 
> * Using rules like awful.rules to define when to transfer ownership
> * Having some ways to running these rules both when the seletion changes and "at any time" like before SIGSTOPping clients

I thought about using it instead of `xclip` to transfer ownership before SIGSTOPping a client, and thought it might be good to see if it needs to be transfered (owner).
